File: client/lib/policy.dart
/// Decoded server policy, used only for presentation and early input hints.
/// Go validates every write independently of these values.
class AppPolicy {
  final List<String> categories, imageExtensions, videoExtensions;
  final int postCharacters,
      messageCharacters,
      campusCharacters,
      aliasCharacters;
  final int maxAttachments, maxImageBytes, maxVideoBytes, maxTotalBytes;

  AppPolicy.fromJson(Map<String, dynamic> json)
    : categories = _strings(json['categories']),
      imageExtensions = _strings(json['media']['imageExtensions']),
      videoExtensions = _strings(json['media']['videoExtensions']),
      postCharacters = _positive(json['limits']['postCharacters']),
      messageCharacters = _positive(json['limits']['messageCharacters']),
      campusCharacters = _positive(json['limits']['campusCharacters']),
      aliasCharacters = _positive(json['limits']['aliasCharacters']),
      maxAttachments = _positive(json['media']['maxAttachments']),
      maxImageBytes = _positive(json['media']['maxImageBytes']),
      maxVideoBytes = _positive(json['media']['maxVideoBytes']),
      maxTotalBytes = _positive(json['media']['maxTotalBytes']);

  static int _positive(dynamic value) {
    if (value is! int || value <= 0) {
      throw const FormatException('Invalid limit');
    }
    return value;
  }

  static List<String> _strings(dynamic value) {
    if (value is! List ||
        value.isEmpty ||
        value.any((v) => v is! String || v.isEmpty)) {
      throw const FormatException('Invalid policy list');
    }
    return List<String>.unmodifiable(value);
  }

  String get formatHint =>
      '${imageExtensions.join(" / ").toUpperCase()} 图片、${videoExtensions.join(" / ").toUpperCase()} 视频';
}

File: server/policy.go
package main

import "net/http"

// These definitions are shared by public input hints and authoritative writes.
const (
	maxPostCharacters    = 1000
	maxMessageCharacters = 2000
	maxCampusCharacters  = 80
	maxAliasCharacters   = 40
	maxAttachments       = 4
	maxImageBytes        = int64(10 << 20)
	maxVideoBytes        = int64(50 << 20)
	maxTotalBytes        = int64(50 << 20)
)

var postCategories = []string{"校园日常", "心事树洞", "搭子集合", "恋爱碎碎念", "学习交流"}

type mediaFormat struct {
	kind, mime string
	extensions []string
}

var mediaFormats = []mediaFormat{
	{"image", "image/jpeg", []string{"jpg", "jpeg"}},
	{"image", "image/png", []string{"png"}},
	{"image", "image/webp", []string{"webp"}},
	{"video", "video/mp4", []string{"mp4"}},
	{"video", "video/webm", []string{"webm"}},
}

func detectMediaFormat(data []byte) mediaFormat {
	mime := http.DetectContentType(data)
	for _, format := range mediaFormats {
		if format.kind == "image" && format.mime == mime ||
			format.mime == "video/mp4" && isMP4(data) ||
			format.mime == "video/webm" && isWebM(data) {
			return format
		}
	}
	return mediaFormat{}
}

func (s *Server) policy(w http.ResponseWriter, r *http.Request) {
	images, videos := []string{}, []string{}
	for _, format := range mediaFormats {
		if format.kind == "image" {
			images = append(images, format.extensions...)
		} else {
			videos = append(videos, format.extensions...)
		}
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"categories": postCategories,
		"limits":     map[string]int{"postCharacters": maxPostCharacters, "messageCharacters": maxMessageCharacters, "campusCharacters": maxCampusCharacters, "aliasCharacters": maxAliasCharacters},
		"media":      map[string]any{"maxAttachments": maxAttachments, "maxImageBytes": maxImageBytes, "maxVideoBytes": maxVideoBytes, "maxTotalBytes": maxTotalBytes, "imageExtensions": images, "videoExtensions": videos},
	})
}
